### ISSUE TITLE
Response metadata

### ISSUE THREAD

**Issue by samselikoff** [samselikoff/ember-cli-mirage#956]
We need to come up with an API to attach metadata (page numbers, sort directions, collection size etc.) to a response.

**Comment by samselikoff** [comment]
There's been some discussions about whether to do this in route handlers or serializers. The pragmatist in me says, why not both.

 - Let's add the ability to attach metadata in route handlers - this is probably where meta specific to the actual return object (model/collection) would be attached (like page info).

 - But also add a method to serializers, which probably has access to schema/db (and possibly the response object), for metadata that is common across multiple routes.

**Comment by Azdaroth** [comment]
I think a pretty convenient way would be to have metadata directly in the collection (primaryResource) and letting the serializer handle it, like here: https://github.com/samselikoff/ember-cli-mirage/pull/916/files#diff-21df86ecb5f80627fb6dc6057c305faaR49. That way we can attach some metadata in route handlers (actually we can do it wherever, as long as we have access to `primaryResource` and it happens before returning final serialized body). It definitely works for pagination (check the PR above), should also work for sort directions, filtering etc.

Not sure about extra metadata being attached directly in serializers, I mean it sounds fine as an idea, but is it really needed? Any particular use case you have in mind?

**Comment by ashrafhasson** [comment]
Although this was discussed in #916 I wanted to make sure this is also kept in mind here. I find that it's important to give users the ability to control how data is paged, e.g., if query params were present and the returned payload is expected to conform to those query params, then one should be able to define that logic in mirage somewhere. Since query params are received in the route handler, this is where I thought one can overload the paginator, but otherwise, how would you alter what the paginator returns. Maybe it needs to be made a service that's injected into the route (if this makes sense)?

**Comment by Azdaroth** [comment]
@ashrafhasson Yeah, that would definitely make sense to have the ability to modify the default behaviour from the route handler. I should find some time the following week to improve it.

**Comment by lypborges** [comment]
I liked the options to handle in route handler and serializers. 
I kwon that the ideia here is to solve the pagination using metadata. What about use the link tag too. http://jsonapi.org/format/#fetching-pagination . This could be a default behaviour  for the default json serializer?

**Comment by hbrysiewicz** [comment]
Feature requests are being moved to Canny.
[View this feature request in Canny](https://ember-cli-mirage.canny.io/feature-requests/p/response-metadata)
